File: android/core/src/main/kotlin/com/openmausbot/companion/core/WebhookMessageContent.kt
package com.openmausbot.companion.core

/** Display projection only. Stored prompts keep their model-facing trust boundaries. */
data class WebhookMessageContent(val task: String, val payload: String?) {
    companion object {
        private val taskMarkers = listOf(
            "AUTHENTICATED WEBHOOK TASK", "USER-CONFIGURED WEBHOOK INSTRUCTIONS", "DEFAULT WEBHOOK INSTRUCTIONS",
        )

        /** Matches the desktop webhook-message parser. Incomplete envelopes remain ordinary text. */
        fun parse(text: String): WebhookMessageContent? {
            val eventStart = text.indexOf("[UNTRUSTED WEBHOOK EVENT DATA]\n").takeIf { it >= 0 } ?: return null
            val trustedPrefix = text.substring(0, eventStart)
            fun block(marker: String, source: String): String? {
                val opening = "[$marker]\n"
                val start = source.indexOf(opening).takeIf { it >= 0 } ?: return null
                val contentStart = start + opening.length
                val end = source.indexOf("\n[/$marker]", contentStart).takeIf { it >= 0 } ?: return null
                return source.substring(contentStart, end)
            }
            val task = taskMarkers.firstNotNullOfOrNull { block(it, trustedPrefix)?.trim()?.takeIf(String::isNotEmpty) } ?: return null
            val event = block("UNTRUSTED WEBHOOK EVENT DATA", text)?.takeIf(String::isNotEmpty) ?: return null
            val splitAt = event.indexOf("\n\n")
            val payload = if (splitAt >= 0) event.substring(splitAt + 2).trim().takeIf(String::isNotEmpty) else null
            return WebhookMessageContent(task, payload)
        }
    }
}

val Message.webhookContent: WebhookMessageContent?
    get() = if (role == Message.Role.USER && kind == Message.Kind.TEXT) text?.let(WebhookMessageContent::parse) else null

File: ios/Sources/CompanionCore/WebhookMessageContent.swift
import Foundation

/// A display projection only: the stored prompt retains its trust wrappers.
/// Matches the desktop's webhook-message parser.
public struct WebhookMessageContent: Hashable, Sendable {
    public let task: String
    public let payload: String?

    public static func parse(_ text: String) -> Self? {
        guard let eventStart = text.range(of: "[UNTRUSTED WEBHOOK EVENT DATA]\n") else { return nil }
        let trustedPrefix = String(text[..<eventStart.lowerBound])
        func block(_ marker: String, in source: String) -> String? {
            guard let start = source.range(of: "[\(marker)]\n"),
                  let end = source.range(of: "\n[/\(marker)]", range: start.upperBound..<source.endIndex) else { return nil }
            return String(source[start.upperBound..<end.lowerBound])
        }
        let markers = ["AUTHENTICATED WEBHOOK TASK", "USER-CONFIGURED WEBHOOK INSTRUCTIONS", "DEFAULT WEBHOOK INSTRUCTIONS"]
        guard let task = markers.compactMap({ block($0, in: trustedPrefix)?.trimmingCharacters(in: .whitespacesAndNewlines) })
            .first(where: { !$0.isEmpty }),
              let event = block("UNTRUSTED WEBHOOK EVENT DATA", in: text), !event.isEmpty else { return nil }
        let payload = event.range(of: "\n\n").map {
            String(event[$0.upperBound...]).trimmingCharacters(in: .whitespacesAndNewlines)
        }
        return Self(task: task, payload: payload?.isEmpty == false ? payload : nil)
    }
}

public extension Message {
    var webhookContent: WebhookMessageContent? {
        guard role == .user, kind == .text, let text else { return nil }
        return WebhookMessageContent.parse(text)
    }
}
